---
title: "Selecting markers for marker-assisted selection in maize"
author: Lindsay Clark, HPCBio, Roy J. Carver Biotechnology Center, University of Illinois,
  Urbana-Champaign
date: "April 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(Biostrings.coloring = FALSE)
```

This document is modified from the
[marker-assisted selection](Marker_assisted_selection.md) workflow.

## Setup

Be sure you have completed the [setup](Setup.md) steps before working through
this pipeline.  First we'll load all the needed packages and functions.

```{r libs, message = FALSE, warning = FALSE}
library(VariantAnnotation)
library(Rsamtools)
library(ggplot2)
library(viridis)
source("src/marker_stats.R")
source("src/getNumGeno.R")
source("src/qtl_markers.R")
```

### Reference genome

Here we'll load the reference genome sequence for maize.  This was
already indexed using `indexFa` (see the setup steps).

The genome was obtained from https://download.maizegdb.org/Zm-B73-REFERENCE-GRAMENE-4.0/

```{r refgenome}
refgenome <- FaFile("data/Zm-B73-REFERENCE-GRAMENE-4.0.fasta")
```

### SNP positions from sequence

A spreadsheet was provided listing the markers of interest and their GBS tags.
Previous analysis indicated that many SNP positions were incorrect, so we will
attempt to determine the correct positions by sequence alignment.

```{r readqtl}
maize_qtl <- read.csv("data/Markers identified from QTL or GWAS studies with sequences - Latest - LVC edits.csv",
                      stringsAsFactors = FALSE)
```

We'll make a consensus sequence for each marker.

```{r consensus}
maize_qtl$Seq.consensus <- sapply(seq_len(nrow(maize_qtl)),
                                  function(i){
                                    consensusString(c(maize_qtl$Allele.1[i], maize_qtl$Allele.2[i]), ambiguityMap = IUPAC_CODE_MAP, threshold = 0.25)
                                  })
```


```{r exportfasta}
allele_seq <- DNAStringSet(maize_qtl$Seq.consensus)
names(allele_seq) <- paste(maize_qtl$QTL.marker, maize_qtl$Study, maize_qtl$Traits, sep = "; ")
writeXStringSet(allele_seq, filepath = "data/maize_qtl_sequence_2021-04-05.fasta")
```

#### BLAST

Alignment was performed with BLAST+/2.10.1, run on the Biocluster at the University of Illinois.

```{bash}
makeblastdb -in Zm-B73-REFERENCE-GRAMENE-4.0.fa -out blast/Zm-B73-4.0 -dbtype nucl -parse_seqids \
  -title Zm-B73-REFERENCE-GRAMENE-4.0

blastn -query data/maize_qtl_sequence_2021-04-05.fasta -db data/reference/blast/Zm-B73-4.0 \
  -num_threads 2 -outfmt 7 -out results/blast/maize_qtl_sequence_2021-04-05.blast.txt
```

How many BLAST hits did we have for each marker?

```{r blastfilter}
blast_lines <- readLines("results/maize_qtl_sequence_2021-04-05.blast.txt")
maize_qtl$N_blast_hits <- as.integer(sub("# ", "", sub(" hits found", "", grep("hits found", blast_lines, value = TRUE))))
```

We will identify the SNP position for each top hit where possible.

```{r eval = FALSE}
query_lines <- grep("^# Query:", blast_lines, value = TRUE)
tophit_lines <- blast_lines[grep("hits found", blast_lines) + 1]
tophit_lines[startsWith(tophit_lines, "#")] <- NA
cat(tophit_lines[!is.na(tophit_lines)], file = "results/maize_qtl_sequence_2021-04-05_tophits.blast.txt", sep = "\n")
```

```{r readblast}
maize_blast <- read.delim("results/maize_qtl_sequence_2021-04-05_tophits.blast.txt", header = FALSE, comment.char = "#")
colnames(maize_blast) <- c("query", "subject", "pct.identity", "alignment.length", "mismatches", "gap.opens",
                           "q.start", "q.end", "s.start", "s.end", "evalue", "bit.score")
maize_blast$query <- sub(";", "", maize_blast$query)
```

```{r}
maize_qtl$Blast_chrom[!is.na(tophit_lines)] <- maize_blast$subject
maize_qtl$Blast_pos[!is.na(tophit_lines)] <- 
  sapply(seq_len(nrow(maize_blast)),
         function(i){
           if(maize_blast$gap.opens[i] > 0) return(NA)
           snp_pad <- maize_qtl$SNP.Position[!is.na(tophit_lines)][i]
           snp_pad <- as.integer(sub(":.*$", "", snp_pad))
           if(maize_blast$s.end[i] > maize_blast$s.start[i]){
             return(maize_blast$s.start[i] + snp_pad - maize_blast$q.start[i] + 1L)
           } else {
             return(maize_blast$s.start[i] - snp_pad + maize_blast$q.start[i] - 1L)
           }
         })
maize_qtl$Blast_strand[!is.na(tophit_lines)] <-
  ifelse(maize_blast$s.end > maize_blast$s.start, "top", "bot")
```

For a few of them, SNP position could not be determined due to gap openings.  Those
sequences I BLASTed again with a different output format to determine the SNP positions.

```{r}
maize_qtl[!is.na(maize_qtl$Blast_chrom) & is.na(maize_qtl$Blast_pos), c("QTL.marker", "Seq.consensus")]

maize_qtl$Blast_pos[match("S10_138511949", maize_qtl$QTL.marker)] <- 138511994L
maize_qtl$Blast_pos[maize_qtl$QTL.marker == "S7_2840287"] <- 28402383L
maize_qtl$Blast_pos[match("S2_227406057", maize_qtl$QTL.marker)] <- 227406100L
maize_qtl$Blast_pos[maize_qtl$QTL.marker == "S8_164863938"] <- 164863937L # The SNP is an insertion

```


View how well the BLAST results match the provided data.  Some positions match exactly,
and most others are fairly close.

```{r}
maize_qtl[,c("QTL.marker", "Chrom", "Chrom.Pos", "Blast_chrom", "Blast_pos")]

temp <- maize_qtl$Chrom.Pos - maize_qtl$Blast_pos
hist(temp)
hist(temp[temp > -1.5e6])

maize_qtl[which(temp < -1.5e6), c("QTL.marker", "Chrom", "Chrom.Pos", "Blast_chrom", "Blast_pos")]
```

One had the wrong chromosome, so I'll manually correct it.

```{r}
maize_qtl$Blast_pos[match("S1_51819212", maize_qtl$QTL.marker)] <- 51819243L
maize_qtl$Blast_chrom[match("S1_51819212", maize_qtl$QTL.marker)] <- "Chr1"
maize_qtl$Blast_strand[match("S1_51819212", maize_qtl$QTL.marker)] <- "bot"
```

#### Bowtie2

A few markers did not yield BLAST results, so we'll see if we can get an alignment with Bowtie2.
Version 2.4.2 was used on the Biocluster at the University of Illinois.

```{bash}
bowtie2-build --threads 4 Zm-B73-REFERENCE-GRAMENE-4.0.fa bowtie/Zm-B73-4.0

bowtie2 --threads 2 -f -x data/reference/bowtie/Zm-B73-4.0 -U data/maize_qtl_sequence_2021-04-05.fasta \
  -S results/bowtie/maize_qtl_sequence_2021-04-05_bowtie_verysensitive.sam --very-sensitive
```

```{r}
samlines <- readLines("results/maize_qtl_sequence_2021-04-05_bowtie_verysensitive.sam")
samlines <- samlines[!startsWith(samlines, "@")]
samsplit <- strsplit(samlines, split = "\t")

maize_qtl$Bowtie_chrom <- sapply(samsplit,
                                 function(x){
                                   out <- x[3]
                                   if(out == "*") return(NA) else return(out)
                                 })

snppos_sam <- function(snp_pad, pos, flag, cigar){
  if(flag == "4") return(NA)
  pos <- as.integer(pos)
  cig_nums <- as.integer(strsplit(cigar, split = "[MID]")[[1]])
  cig_char <- strsplit(cigar, split = "[[:digit:]]")[[1]]
  cig_char <- cig_char[cig_char != ""]
  if(flag == "16"){
    snp_pad <- sum(cig_nums[cig_char %in% c("M", "I")]) - snp_pad - 1L
  }
  out <- pos
  for(i in seq_along(cig_nums)){
    if(cig_char[i] == "M"){
      if(cig_nums[i] > snp_pad){
        out <- out + snp_pad
        break
      } else {
        out <- out + cig_nums[i]
        snp_pad <- snp_pad - cig_nums[i]
      }
    }
    if(cig_char[i] == "I"){
      snp_pad <- snp_pad - cig_nums[i]
    }
    if(cig_char[i] == "D"){
      out <- out + cig_nums[i]
    }
  }
  return(out)
}

maize_qtl$Bowtie_pos <- sapply(seq_len(nrow(maize_qtl)),
                               function(i){
                                 snp_pad <- as.integer(sub(":.*$", "", maize_qtl$SNP.Position[i]))
                                 snppos_sam(snp_pad, samsplit[[i]][4], samsplit[[i]][2],
                                            samsplit[[i]][6])
                               })

maize_qtl$Bowtie_strand <- sapply(samsplit,
                                  function(x){
                                    if(x[2] == "0") return("top")
                                    if(x[2] == "16") return("bot")
                                    if(x[2] == "4") return(NA_character_)
                                  })
```

The only places where Bowtie2 found a different position from BLAST were cases
where it aligned to an entirely different chromosome.

```{r}
maize_qtl[which(maize_qtl$Blast_pos != maize_qtl$Bowtie_pos),
          c("QTL.marker", "Chrom", "Chrom.Pos", "Blast_chrom", "Blast_pos", "Bowtie_chrom", "Bowtie_pos")]
```

Where no BLAST alignment was found, sometimes Bowtie2 found an alignment,
although it did not always match the chromosome reported by the genotyping company.

```{r}
maize_qtl[is.na(maize_qtl$Blast_pos),
          c("QTL.marker", "Chrom", "Chrom.Pos", "Blast_chrom", "Blast_pos", "Bowtie_chrom", "Bowtie_pos")]
```

Let's start columns to store the final chromosome and position that we will use
after comparing both alignments.

```{r}
maize_qtl$Final_chrom <- maize_qtl$Blast_chrom
maize_qtl$Final_pos <- maize_qtl$Blast_pos
maize_qtl$Final_strand <- maize_qtl$Blast_strand
bad_blast <- which(is.na(maize_qtl$Blast_chrom) |
                     (maize_qtl$Blast_chrom != paste0("Chr", maize_qtl$Chrom) &
                        maize_qtl$Blast_pos != maize_qtl$Chrom.Pos))
bad_blast
maize_qtl$Final_chrom[bad_blast] <- NA
maize_qtl$Final_pos[bad_blast] <- NA
maize_qtl$Final_strand[bad_blast] <- NA
good_bowtie <- which(!is.na(maize_qtl$Bowtie_chrom) &
                       (maize_qtl$Bowtie_chrom == paste0("Chr", maize_qtl$Chrom) |
                        maize_qtl$Bowtie_pos == maize_qtl$Chrom.Pos))
bowtie_fill <- good_bowtie[good_bowtie %in% bad_blast]

maize_qtl$Final_chrom[bowtie_fill] <- maize_qtl$Bowtie_chrom[bowtie_fill]
maize_qtl$Final_pos[bowtie_fill] <- maize_qtl$Bowtie_pos[bowtie_fill]
maize_qtl$Final_strand[bowtie_fill] <- maize_qtl$Bowtie_strand[bowtie_fill]
```

#### Pairwise alignments based on supplied positions

Below is a list of markers where the chromosomes identified by neither BLAST
nor Bowtie match the supplied chromosome.  We'll try one more approach to align
them.

```{r}
to_fill <- which(is.na(maize_qtl$Final_chrom))
maize_qtl[to_fill,
          c("QTL.marker", "Chrom", "Chrom.Pos", "Blast_chrom", "Blast_pos", "Bowtie_chrom", "Bowtie_pos", "Seq.consensus")]
```

Given that other SNPs were off in position by up to 80 nt, and the tags
are 69 nt long, we will search sequence 200 nt on either side of each SNP
position.

```{r}
remaining_snps <- GRanges(paste0("Chr", maize_qtl$Chrom[to_fill]),
                          IRanges(start = maize_qtl$Chrom.Pos[to_fill],
                                  width = 1))
names(remaining_snps) <- maize_qtl$QTL.marker[to_fill]
remaining_snps_flank <- flank(remaining_snps, width = 200, both = TRUE)
remaining_snps_flank_seq <- scanFa(refgenome, remaining_snps_flank)
```

We'll do the alignment using the Biostrings package.

```{r}
aligns_forward <- aligns_reverse <- vector(mode = "list", length = 4)
best_align <- character(4)

for(i in 1:4){
  thistag <- DNAString(maize_qtl$Seq.consensus[match(names(remaining_snps)[i], maize_qtl$QTL.marker)])
  aligns_forward[[i]] <- pairwiseAlignment(thistag, remaining_snps_flank_seq[i],
                          substitutionMatrix = nucleotideSubstitutionMatrix(),
                          type = "local", gapOpening = 2, gapExtension = 2)
  aligns_reverse[[i]] <- pairwiseAlignment(reverseComplement(thistag), remaining_snps_flank_seq[i],
                          substitutionMatrix = nucleotideSubstitutionMatrix(),
                          type = "local", gapOpening = 2, gapExtension = 2)
  if(score(aligns_forward[[i]]) > score(aligns_reverse[[i]])){
    best_align[i] <- "top"
  } else {
    best_align[i] <- "bot"
  }
}
```

We can examine the output alignments.

```{r}
best_align

aligns_reverse[[1]]
aligns_forward[[2]]
aligns_reverse[[3]]
aligns_forward[[4]]
```

Now we can calculate SNP positions.

```{r}
maize_qtl$SNP.Position[to_fill]

maize_qtl$Final_pos[to_fill[1]] <- start(remaining_snps_flank)[1] + 188L - 1L + 69L - 15L - 1L
maize_qtl$Final_pos[to_fill[2]] <- start(remaining_snps_flank)[2] + 181L - 1L + 20L
maize_qtl$Final_pos[to_fill[3]] <- start(remaining_snps_flank)[3] + 161L - 1L + 69L - 44L - 1L
maize_qtl$Final_pos[to_fill[4]] <- start(remaining_snps_flank)[4] + 176L - 1L + 25L

maize_qtl$Final_chrom[to_fill] <- as.character(seqnames(remaining_snps))
maize_qtl$Final_strand[to_fill] <- best_align
```

Checking the results:

```{r}
maize_qtl[to_fill,
          c("QTL.marker", "Chrom", "Chrom.Pos", "Final_chrom", "Final_pos")]
```

As with the rest of the markers, some of the positions match and some don't.

Now I will save all of my positions to a CSV file.

```{r eval = FALSE}
write.csv(maize_qtl, file = "results/maize_qtl_positions_corrected.csv")
```

### Public HapMap VCFs for maize

In order to annotate more flanking SNPs than were provided in the original
genotype files,
I downloaded publicly available maize variant data from Cyverse.  See
http://cbsusrv04.tc.cornell.edu/users/panzea/download.aspx?filegroupid=34
for instructions.  Each chromosome is in its own VCF.

```{r bzip_and_index, eval = FALSE}
vcfpath <- "/mnt/lvclark1/maize_snp/" # point to where you have the VCF files
vcffiles <- paste0("hmp321_agpv4_chr", 1:10, ".vcf.gz")
```

I'll import the QTL now so that I can just import the VCF regions of interest
and put them into a smaller file to download to my laptop.  Note that chromosome
names in the reference genome are formatted as "Chr1" whereas chromosome names
in the public VCFs are formatted as "1".

```{r qtl_ranges}


head(maize_qtl)

search_distance <- 1e5
qtl_ranges <- GRanges(maize_qtl$Chrom,
                      IRanges(start = maize_qtl$Position - search_distance,
                              end = maize_qtl$Position + search_distance))
names(qtl_ranges) <- maize_qtl$QTL.marker
```

```{r readbigvcfs, eval = FALSE}
vcflist <- numgenlist <- list()
length(vcflist) <- length(numgenlist) <- 10
for(chr in 1:10){
  qtlsubset <- seqnames(qtl_ranges) == paste0("Chr", chr)
  thisrange <- GRanges(as.character(chr),
                       ranges(qtl_ranges[qtlsubset]))
  names(thisrange) <- names(qtl_ranges[qtlsubset])
  svp <- ScanVcfParam(which = thisrange, geno = NA)
  thisfile <- paste0(vcfpath, vcffiles[chr])
  vcflist[[chr]] <- readVcf(thisfile,
                            genome = as.character(chr),
                            param = svp)
  numgenlist[[chr]] <- getNumGeno(thisfile, thisrange)
}

save(vcflist, numgenlist, file = "results/vcflist.RData")
```

The above code made a file small enough to download to my laptop. I will load
it to continue from here.

```{r loadvcflist}
load("results/vcflist.RData")
```

We should do QC to make sure the reference genome sequence matches what is
listed in the VCF. We'll check chromosome 1.

```{r}
ranges1 <- rowRanges_correctedSeqnames(vcflist[[1]],
                                       fixfn = function(x){
                                         paste0("Chr", x)
                                       })
ranges1

refcheck1 <- scanFa(refgenome, ranges1)
refcheck1

mean(refcheck1 == ranges1$REF) # should be 1
```

Are the SNPs of interest in the public VCF?  If so, we don't need the study-specific
VCFs.

```{r}
maize_qtl$Public <- FALSE
maize_qtl$PublicV3 <- FALSE
for(i in seq_len(nrow(maize_qtl))){
  #qtl <- maize_qtl$QTL.marker[i] # could match to paramRangeID
  chrom <- as.integer(sub("Chr", "", maize_qtl$Chrom[i]))
  pos <- maize_qtl$Position[i]
  maize_qtl$Public[i] <- pos %in% start(vcflist[[chrom]])
  maize_qtl$PublicV3[i] <- paste(chrom, pos, sep = "-") %in% rownames(vcflist[[chrom]])
}

mean(maize_qtl$Public)
mean(maize_qtl$PublicV3)
```

No, the majority are not in the public VCFs.

### Study-specific VCFs

Below are names of the VCFs from the GWAS and QTL studies.  We'll zip and
index them for reading.

```{r files, eval = FALSE}
small_vcfs <- paste0("data/", c("GWAS", "EEWQTL_new", "EEYQTL_new", "EWQTL"), ".vcf")
for(v in small_vcfs){
  b <- bgzip(v)
  indexTabix(b, format = "vcf")
}
```

```{r}
bgzs <- paste0("data/", c("GWAS", "EEWQTL_new", "EEYQTL_new", "EWQTL"), ".vcf.bgz")
names(bgzs) <- c("GWAS", "EEWQTL", "EEYQTL", "EWQTL")
```

```{r}
small_vcf_list <- sapply(bgzs, function(x) readVcf(x, param = ScanVcfParam(geno = NA)))
```

```{r}
ranges2 <- sapply(small_vcf_list, 
                  function(x) rowRanges_correctedSeqnames(x,
                                       fixfn = function(x){
                                         paste0("Chr", x)
                                       }))
# subset chromosomes because there is some bug
# ranges2 <- sapply(ranges2,
#                   function(x) x[seqnames(x) %in% paste0("chr", 1:9)])

refcheck2 <- sapply(ranges2, function(x) scanFa(refgenome, x))

mean(refcheck2[[1]] == ranges2[[1]]$REF | refcheck2[[1]] == unlist(ranges2[[1]]$ALT))
mean(refcheck2[[2]] == ranges2[[2]]$REF | refcheck2[[2]] == unlist(ranges2[[2]]$ALT))
mean(refcheck2[[3]] == ranges2[[3]]$REF | refcheck2[[3]] == unlist(ranges2[[3]]$ALT))
mean(refcheck2[[4]] == ranges2[[4]]$REF | refcheck2[[4]] == unlist(ranges2[[4]]$ALT))

ranges2[[2]][!(refcheck2[[2]] == ranges2[[2]]$REF | refcheck2[[2]] == unlist(ranges2[[2]]$ALT))]
```

```{r}
test <- sub(".*-", "", names(ranges2[[1]]))
test <- as.integer(sub(":[ACGT]>[ACGT]$", "", test))
hist(test)
summary(test)

ranges2a <- sapply(ranges2[c(1, 3, 4)],
                   function(x){
                     pad <- as.integer(sub(":[ACGT]>[ACGT]$", "", sub(".*-", "", names(x))))
                     GRanges(seqnames = seqnames(x),
                             IRanges(start = start(x) + pad - 0,
                                     end = start(x) + pad - 0))
                   }) # 
refcheck2a <- sapply(ranges2a, function(x) scanFa(refgenome, x))

ranges2b <- sapply(ranges2[c(1, 3, 4)],
                   function(x){
                     pad <- as.integer(sub(":[ACGT]>[ACGT]$", "", sub(".*-", "", names(x))))
                     GRanges(seqnames = seqnames(x),
                             IRanges(start = start(x) - pad + 68,
                                     end = start(x) - pad + 68))
                   })
refcheck2b <- sapply(ranges2b, function(x) scanFa(refgenome, x))

mean(refcheck2a[[1]] == ranges2[[1]]$REF |
       refcheck2b[[1]] == ranges2[[1]]$REF)
```


## Data import

### Significant SNPs

We will import a spreadsheet listing markers of interest.  I reformatted the
Excel file that was provided to make it more compatible with R. (I.e., deleted
all header rows aside from the top one, deleted empty rows, merged multiple
rows belonging to the same marker, and listed the trait in each row.)

```{r qtl}
yam_qtl <- read.csv("data/yam_qtl.csv", stringsAsFactors = FALSE)

str(yam_qtl)
```

We will make a chromosome column to match the chromosome names in the FASTA and
VCF files.  We'll also make a marker name column with the allele trimmed off.

```{r qtlchr}
yam_qtl$Chromosome <- sub("_[[:digit:]]+_[ACGT]$", "", yam_qtl$Marker)
yam_qtl$Marker_short <- sub("_[ACGT]$", "", yam_qtl$Marker)
head(yam_qtl)
```

### Phenotypes

We can read in phenotype data so that we can see how well SNPs predict phenotypes.

```{r importpheno}
pheno <- read.csv("data/pheno_data all 174.csv")
head(pheno)
```

The column names should be made to match the QTL spreadsheet.

```{r matchtraits}
traits <- unique(yam_qtl$Trait)
names(pheno)

names(pheno) <- gsub("_201[78]", "", names(pheno))
names(pheno) <- gsub("_", " ", names(pheno))
names(pheno)

setdiff(traits, names(pheno)) # traits from the QTL file that haven't been matched in the phenotype file
setdiff(names(pheno), traits) # traits from the phenotype file that haven't been matched in the QTL file

names(pheno)[names(pheno) == "Spines on tuber"] <- "Spines on tuber surface"
names(pheno)[names(pheno) == "No of tubers"] <- "Number of tubers per plant"
names(pheno)[names(pheno) == "Yield plant"] <- "Yield per plant"

all(traits %in% names(pheno)) # should be TRUE
```

### Genotypes and annotations from VCF

We will specify ranges in which we wish to look at SNPs for KASP marker design.
Let's look within 100 kb of each significant SNP.

```{r qtlranges}

```

We will import numeric genotypes just within these ranges.

```{r numgen}
numgen <- getNumGeno(bg, ranges = qtl_ranges)
str(numgen)
```

There are 5684 markers across 173 individuals, and genotypes are coded from
zero to two.  We will change the accession names to match the phenotype spreadsheet.

```{r matchaccessions}
if(all(sub("_", "", colnames(numgen)) %in% pheno$DRS)){
  colnames(numgen) <- sub("_", "", colnames(numgen))
}

all(colnames(numgen) %in% pheno$DRS) # should be TRUE
```

We will also import SNP metadata within these ranges.

```{r importvcf}
myvcf <- readVcf(bg,
                 param = ScanVcfParam(geno = NA, which = qtl_ranges))

rowRanges(myvcf)
```

We can see that the `paramRangeID` column indicates which original marker each
SNP is near.  Since there were some significant SNPs close to each other, that
also means we have some duplicates in both `numgen` and `myvcf`.

```{r dupcheck}
identical(rownames(numgen), names(rowRanges(myvcf)))

as.logical(anyDuplicated(rownames(numgen)))
```

### Unfiltered VCF

In this case we had a much larger VCF with rarer SNPs, so we will import that too.

```{r imporvcf2}
bigvcf <- readRDS(rds)

rowRanges(bigvcf)
```

Since we have quality scores, we will look at the distribution.

```{r qualhist}
hist(rowRanges(bigvcf)$QUAL, xlab = "Quality score",
     main = "Histogram of quality scores in large VCF")
```

This suggests filtering to only keep the highest scores is advisable.
We will also make sure to keep any SNPs that were in our smaller VCF.

```{r filtervcf}
temp <- paste(seqnames(bigvcf), start(bigvcf), sep = "_")

bigvcf <- bigvcf[rowRanges(bigvcf)$QUAL > 900 | 
                temp %in% names(myvcf),]
rm(temp)
```

Lastly, we will filter the VCF to only contain SNPs in our QTL ranges.

```{r subsetvcf}
bigvcf <- subsetByOverlaps(bigvcf, qtl_ranges)
```

## Technical parameters for marker design from significant SNPs

Ideally, we would like to design markers directly from the significant hits.
We should check that they will make reasonably good KASP markers first,
however.

### GC content

PCR will work best when GC content is 40-60%.

```{r gccontent}
yam_qtl$GCcontent <- gcContent(myvcf, yam_qtl$Marker_short, refgenome)

hist(yam_qtl$GCcontent, xlab = "GC content",
     main = "GC content flanking significant SNPs", breaks = 20)
```

Many are below the desired range, so we may see if there are any nearby
SNPs in LD that have better GC content.

### Number of flanking SNPs

A few flanking SNPs are okay, but we want to make sure none of these
have an excessive amount.

```{r flankingsnps}
yam_qtl$Nflanking <- nFlankingSNPs(bigvcf, yam_qtl$Marker_short)

hist(yam_qtl$Nflanking, xlab = "Number of flanking SNPs",
     main = "Number of flanking SNPs within 50 bp of significant SNPs")
table(yam_qtl$Nflanking)
```

For those with three or more, we might see if there are better markers.

## Evaluating nearby markers

### Finding markers in linkage disequilibrium (LD)

Below is a function that uses that information to estimate LD of every SNP
within range with the significant SNP.  We will reorder the results to match
the table of significant SNPs.

```{r ld}
ld <- LD(numgen, myvcf)
ld <- ld[yam_qtl$Marker_short]
```

Let's also extract start positions for the sake of plotting.

```{r grangeslist}
snplist <- split(rowRanges(myvcf), rowRanges(myvcf)$paramRangeID)
snplist <- snplist[yam_qtl$Marker_short]
positions <- start(snplist)
```

```{r plotld}
i <- 1
ggplot(mapping = aes(x = positions[[i]], y = ld[[i]])) +
  geom_point(alpha = 0.3) +
  labs(x = "Position", y = "R-squared",
       title = paste("Linkage disequilibrium with", names(snplist)[i]))
```

The actual SNP of interest shows up as 100% LD in the middle of the range.  There
are a few nearby around 50% LD, which is not great but we might consider those if
the SNP of interest is in a very low GC region, for example.

### Finding markers correlating with the trait

For all markers nearby to our significant SNPs, let's also look at the R-squared
with the corresponding trait.

```{r rsq, warning = FALSE}
phen_rsq <- phenoCorr(numgen, myvcf, yam_qtl$Marker_short, yam_qtl$Trait,
                      pheno) ^ 2
phen_rsq <- phen_rsq[yam_qtl$Marker_short]
```

```{r plotrsq}
ggplot(mapping = aes(x = positions[[i]], y = phen_rsq[[i]],
                     color = ld[[i]])) +
  geom_point(alpha = 0.7) +
  labs(x = "Position", y = "R-squared",
       title = paste("Association with", yam_qtl$Trait[i]),
       color = "LD with hit") +
  scale_color_viridis()
```

## Choosing markers to output

We may want to output a table of the best markers, along with some statistics so that
we can manually choose among them.  Let's take the top ten R-squared values for
association with the trait of interest, and also make sure to get the significant
SNP itself.

```{r top10}
n <- 10 # edit this number if you want to keep a different number of markers

top10 <- lapply(phen_rsq, function(x){
  x1 <- sort(x, decreasing = TRUE)
  if(length(x1) > n){
    x1 <- x1[1:n]
  }
  return(names(x1))
} )

top10tab <- utils::stack(top10)
colnames(top10tab) <- c("SNP_ID", "QTL")

# Add in any QTL that weren't in top 10 associated SNPs
toadd <- setdiff(yam_qtl$Marker_short, top10tab$SNP_ID)
top10tab <- rbind(top10tab,
                  data.frame(SNP_ID = toadd, QTL = toadd))
top10tab <- top10tab[order(factor(top10tab$QTL, levels = yam_qtl$Marker_short)),]
```

Now we'll get the KASP-formatted sequence for these markers.

```{r formatkasp}
outtab <- formatKasp(bigvcf, top10tab$SNP_ID, refgenome)
outtab <- cbind(outtab, top10tab[,"QTL", drop = FALSE])
```

We'll add in linkage disequilibrium and trait association data, and also mark
which allele was positively associated with the trait.

```{r ldrsq, warning = FALSE}
extr <- function(x, y, lst){
  return(lst[[x]][y])
}
outtab$LD_with_QTL <- mapply(extr, outtab$QTL, outtab$SNP_ID, MoreArgs = list(ld))
outtab$R2_with_trait <- mapply(extr, outtab$QTL, outtab$SNP_ID, MoreArgs = list(phen_rsq))

als <- whichAlleleList(numgen, myvcf, yam_qtl$Marker_short, yam_qtl$Trait,
                       pheno)
als <- als[yam_qtl$Marker_short]
outtab$Pos_allele <- mapply(function(x, y, ...) as.character(extr(x, y, ...)),
                            outtab$QTL, outtab$SNP_ID, MoreArgs = list(als))
```

We will also add the GC content and number of flanking SNPs.

```{r gcouttab}
outtab$GC_content <- gcContent(myvcf, outtab$SNP_ID, refgenome)
outtab$N_flanking <- nFlankingSNPs(bigvcf, outtab$SNP_ID)

head(outtab)
```

Now we have a data frame that we can export to a spreadsheet, and we can
manually select SNPs for development as KASP markers.  I recommend selecting
the SNP matching the QTL, plus one or two more SNPs that are similarly
associated with the trait and have as close to optimal GC content as possible.

```{r export}
write.csv(outtab, file = "results/mas_markers.csv", row.names = FALSE)
```

