---
title: "Selecting markers for pedigree verification"
author: Lindsay Clark, HPCBio, Roy J. Carver Biotechnology Center, University of Illinois,
  Urbana-Champaign
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this workflow, we'll identify markers useful for identifying accessions and
verifying pedigrees, and export those markers for KASP marker design.

## Setup

Be sure you have completed the [setup](Setup.md) steps before working through
this pipeline.  First we'll load all the needed packages and functions.

```{r libs, message = FALSE, warning = FALSE}
library(VariantAnnotation)
library(Rsamtools)
library(adegenet)
library(ggplot2)
source("src/marker_stats.R")
source("src/getNumGeno.R")
source("src/evaluate_marker_sets.R")
```

Next we'll point to our VCF and reference genome files. Edit these lines to
point to your dataset.

```{r files}
bg <- "data/331_new_data_vcf_IBRC.vcf.bgz"
refgenome <- FaFile("data/TDr96_F1_v2_PseudoChromosome.rev07.fasta")
```

## Population statistics

We'll import all the genotypes in numeric format, indicating number of copies
of the alternative allele.

```{r numeric_gen}
numgen <- getNumGeno(bg)
numgen[1:10,1:10]
```

Let's look at the distribution of allele frequencies in this dataset.

```{r alfreq}
alfreq <- rowMeans(numgen, na.rm = TRUE) / 2
hist(alfreq, xlab = "Allele frequency", main = "Histogram of alternative allele frequency")
```

We can look at the relationship between observed and expected heterozygosity
( _Ho_ and _He_, respectively).

```{r hohe}
He <- Expected_het(alfreq)
Ho <- rowMeans(numgen == 1, na.rm = TRUE)
hist(Ho / He)
```

In this case the mode is a little above 1, suggesting the presence of hybrid
lines and/or polyploids in the dataset.  The bump at 2 indicates likely paralogs.
In this case we'll set a cutoff at 1.5.

```{r hohe_vs_alfreq}
hohe_cutoff <- 1.5

ggplot(mapping = aes(x = alfreq, y = Ho/He)) +
  geom_point(alpha = 0.05) +
  geom_hline(yintercept = hohe_cutoff, color = "blue") +
  labs(x = "Allele frequency") +
  ggtitle("Filtering for putative paralogs")
```

We can see what proportion of markers we would lose at this cutoff,  It's about 11%.

```{r}
mean(Ho/He > hohe_cutoff)
```

We'll do that filtering now, then start a data frame to track statistics about each SNP.

```{r}
keep1 <- which(Ho/He < hohe_cutoff)

numgen <- numgen[keep1,]
snpdf <- data.frame(row.names = rownames(numgen),
                    AlFreq = alfreq[keep1],
                    Ho = Ho[keep1],
                    He = He[keep1])
```

In this case there are no missing data, but I am going to add that calculation
in for the sake of reusing the workflow.

```{r}
snpdf$MissingRate <- rowMeans(is.na(numgen))
summary(snpdf)
```

We can also look at the heterozygosity across individuals.  The two outlier
individuals we see below are likely to be polyploids or inter-species hybrids.
We'll filter them out.

```{r}
hetByInd <- colMeans(numgen == 1, na.rm = TRUE)
hist(hetByInd)

numgen <- numgen[, hetByInd < 0.5]
```

## Genetic groups

We could simply choose markers with high allele frequency at this point, but
instead I am going to identify genetic groups first.  That way, I can test the
allele frequency within each genetic group.  We generally want a variety of
markers so that we have some that are variable within each group, and some
that can distiguish groups.  We'll use the Discriminant Analysis of Principal
Components (DAPC) protocol implemented in the `adegenet` package.  In order
to knit this document I have to set `n.pca`, `n.clust`, and `n.da` manually, but for
your work I recommend a first pass where they are left as `NULL` so that you
can chose them based on where the graphs plateau.

```{r echo = FALSE}
set.seed(1016) # to ensure report reproducibility
```

```{r}
myclust <- find.clusters(t(numgen), n.pca = 100, n.clust = 10)
```

Check the size of the groups.  If it is very uneven, you may want fewer groups.

```{r}
myclust$size
```

Now perform DAPC.

```{r}
mydapc <- dapc(t(numgen), grp = myclust$grp, scale = TRUE, n.pca = 50, n.da = 2)
```

```{r}
scatter(mydapc, posi.da = "topright")
```

Group assignments are in `mydapc$assign`.  You many want to inspect some of the
more differentiated groups to see if they are biologically meaningful.

```{r}
colnames(numgen)[mydapc$assign == 9]
colnames(numgen)[mydapc$assign == 4]
```

Now we can get allele frequencies within each genetic group.

```{r}
grp_alfreq <- sapply(levels(mydapc$assign),
                     function(x) rowMeans(numgen[, mydapc$assign == x], na.rm = TRUE) / 2)
colnames(grp_alfreq) <- paste0("AlFreq_Grp", colnames(grp_alfreq))
head(grp_alfreq)
```

