---
title: "Selecting markers for pedigree verification"
author: Lindsay Clark, HPCBio, Roy J. Carver Biotechnology Center, University of Illinois,
  Urbana-Champaign
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

In this workflow, we'll identify markers useful for identifying accessions and
verifying pedigrees, and export those markers for KASP marker design.

## Setup

Be sure you have completed the [setup](Setup.md) steps before working through
this pipeline.  First we'll load all the needed packages and functions.

```{r libs, message = FALSE, warning = FALSE}
library(VariantAnnotation)
library(Rsamtools)
library(adegenet)
library(ggplot2)
library(ape)
library(dplyr)
source("src/marker_stats.R")
source("src/getNumGeno.R")
source("src/evaluate_marker_sets.R")
```

Next we'll point to our VCF and reference genome files. Edit these lines to
point to your dataset.

```{r files}
bg <- "data/331_new_data_vcf_IBRC.vcf.bgz"
refgenome <- FaFile("data/TDr96_F1_v2_PseudoChromosome.rev07.fasta")
```

## Population statistics

We'll import all the genotypes in numeric format, indicating number of copies
of the alternative allele.

```{r numeric_gen}
numgen <- getNumGeno(bg)
numgen[1:10,1:10]
```

### Check for clonal duplicates

It is good to check that there aren't any clonal duplicates in the dataset, as
these could bias downstream analysis.  We'll calculate inter-individual distances
and look at the distribution of those distances.

```{r echo = FALSE}
set.seed(1019) # to make report reproducible
```

```{r inddist}
mydist <- interIndividualDist(numgen)

hist(mydist, xlab = "Euclidian distance", main = "Histogram of inter-individual distances")
```

```{r njtree, fig.height = 6, fig.width = 6}
plot(nj(mydist), type = "unrooted", show.tip.label = FALSE)
```

Here it looks like the most closely related individuals are family members, not
clones.  If there were some values closer to zero, however, clones could be
removed using the `removeClones` function.

```{r removeclones}
numgen <- removeClones(numgen, mydist)
```

### Observed and expected heterozygosity

Let's look at the distribution of allele frequencies in this dataset.

```{r alfreq}
alfreq <- rowMeans(numgen, na.rm = TRUE) / 2
hist(alfreq, xlab = "Allele frequency", main = "Histogram of alternative allele frequency")
```

We can look at the relationship between observed and expected heterozygosity
( _Ho_ and _He_, respectively).

```{r hohe}
He <- Expected_het(alfreq)
Ho <- rowMeans(numgen == 1, na.rm = TRUE)
hist(Ho / He)
```

In this case the mode is a little above 1, suggesting the presence of hybrid
lines and/or polyploids in the dataset.  The bump at 2 indicates likely paralogs.
In this case we'll set a cutoff at 1.5.

```{r hohe_vs_alfreq}
hohe_cutoff <- 1.5

ggplot(mapping = aes(x = alfreq, y = Ho/He)) +
  geom_point(alpha = 0.05) +
  geom_hline(yintercept = hohe_cutoff, color = "blue") +
  labs(x = "Allele frequency") +
  ggtitle("Filtering for putative paralogs")
```

We can see what proportion of markers we would lose at this cutoff.  It's about 11%.

```{r mean_hohe}
mean(Ho/He > hohe_cutoff)
```

We'll do that filtering now, then start a data frame to track statistics about each SNP.

```{r filter_hohe}
keep1 <- which(Ho/He < hohe_cutoff)

numgen <- numgen[keep1,]
snpdf <- data.frame(row.names = rownames(numgen),
                    AlFreq = alfreq[keep1],
                    Ho = Ho[keep1],
                    He = He[keep1])
```

In this case there are no missing data, but I am going to add that calculation
in for the sake of reusing the workflow.

```{r missrate}
snpdf$MissingRate <- rowMeans(is.na(numgen))
summary(snpdf)
```

We can also look at the heterozygosity across individuals.  The two outlier
individuals we see below are likely to be polyploids or inter-species hybrids.
We'll filter them out.

```{r filter_ind}
hetByInd <- colMeans(numgen == 1, na.rm = TRUE)
hist(hetByInd)

numgen <- numgen[, hetByInd < 0.5]
```

## Genetic groups

We could simply choose markers with high allele frequency at this point, but
instead I am going to identify genetic groups first.  That way, I can test the
allele frequency within each genetic group.  We generally want a variety of
markers so that we have some that are variable within each group, and some
that can distinguish groups.  We'll use the Discriminant Analysis of Principal
Components (DAPC) protocol implemented in the `adegenet` package.  In order
to knit this document I have to set `n.pca`, `n.clust`, and `n.da` manually, but for
your work I recommend a first pass where they are left as `NULL` so that you
can chose them based on where the graphs plateau.

```{r echo = FALSE}
set.seed(1016) # to ensure report reproducibility
```

```{r findclusters}
myclust <- find.clusters(t(numgen), n.pca = 100, n.clust = 10)
```

Check the size of the groups.  If it is very uneven, you may want fewer groups.

```{r clustsize}
myclust$size
```

Now perform DAPC.

```{r dapc}
mydapc <- dapc(t(numgen), grp = myclust$grp, scale = TRUE, n.pca = 50, n.da = 2)
```

```{r dapcscatter}
scatter(mydapc, posi.da = "topright")
```

Group assignments are in `mydapc$assign`.  You many want to inspect some of the
more differentiated groups to see if they are biologically meaningful.

```{r checkgroups}
colnames(numgen)[mydapc$assign == 9]
colnames(numgen)[mydapc$assign == 4]
```

Now we can get allele frequencies within each genetic group.

```{r freq_by_group}
grp_alfreq <- sapply(levels(mydapc$assign),
                     function(x) rowMeans(numgen[, mydapc$assign == x], na.rm = TRUE) / 2)
colnames(grp_alfreq) <- paste0("AlFreq_Grp", colnames(grp_alfreq))
head(grp_alfreq)
```

## Technical parameters

For this portion of the workflow, we'll import SNP metadata using the
`VariantAnnotation` package.

```{r importvcf}
myvcf <- readVcf(bg,
                 param = ScanVcfParam(geno = NA))

rowRanges(myvcf)
```

### GC content

PCR tends to work best in regions that are 40-60% GC.  We will test
the GC content for the 50-bp flanking region for each SNP in our
table.

```{r gccontent}
snpdf$GCcontent <- gcContent(myvcf, rownames(snpdf), refgenome)

hist(snpdf$GCcontent, xlab = "GC content", main = "Histogram of GC content")

mean(snpdf$GCcontent >= 0.4 & snpdf$GCcontent <= 0.6)
```

About 36.5% of markers are within the target range for GC content.
We can subset to just look at these.

```{r gcsubset}
snpdf2 <- filter(snpdf, GCcontent >= 0.4, GCcontent <= 0.6)
```

### Number of flanking SNPs

Although we'll annotate flanking SNPs, the fewer there are, the more likely
the marker won't have technical problems.

```{r nflank}
snpdf2$Nflanking <- nFlankingSNPs(myvcf, rownames(snpdf2))

hist(snpdf2$Nflanking, xlab = "Number of flanking SNPs",
     main = "Histogram of number of flanking SNPs")
```

There seem to be plenty with no flanking SNPs, so we will just keep those.

```{r flankfilt}
snpdf3 <- filter(snpdf2, Nflanking == 0)
```

## Choosing a set of markers

Since we are selecting a small set of markers, we want to not only make sure
that each individual marker has a high minor allele frequency to maximize its
information content, but also that the markers complement each other well so
that they can distinguish individuals and genetic groups.

Two approaches are presented here.  One is a Galaxy tool developed by Carlos
Ignacio for finding sets of markers that distinguish all accessions.  The
second is an algorithm I developed to maximize diversity captured within and
between genetic groups.

### Galaxy tool

First, we have to export our set of suitable markers back to VCF.

```{r subvcf, eval = FALSE}
rr3 <- rowRanges(myvcf)[rownames(snpdf3)]
subvcf <- readVcf(bg, genome = seqinfo(myvcf),
                  param = ScanVcfParam(which = rr3))

writeVcf(subvcf, filename = "results/marker_subset.vcf")
```

Now, on Galaxy, use the
[Purity tool](http://galaxy-demo.excellenceinbreeding.org/root?tool_id=purity_beta)
on that VCF file.  I set it to select 50 markers after considering 10,000
solutions, and left the distance cutoff at 0.05.  Two files were output, and I
downloaded the second to my `results` folder.  Below I will import it to get
the marker set.

```{r}
markers_purity <- read.delim("results/Galaxy6-[Purity_(beta)_on_data_5].tabular")$Name[1:50]

sort(markers_purity)
```

### Simulated annealing algorithm

In the file [evaluate_marker_sets.R](src/evaluate_marker_sets.R) there is
a function to use a
[simulated annealing algorithm](https://en.wikipedia.org/wiki/Simulated_annealing)
to find optimal sets of markers that capture diversity within and between
genetic groups.

```{r echo = FALSE}
set.seed(1020) # for report reproducibility
```

```{r choosemarkers, message = FALSE}
markers_simanneal <- findMarkerSet(grp_alfreq[rownames(snpdf3),], nSNP = 50)$Set

sort(markers_simanneal)
```

### Comparing the results

How well do the markers distinguish individuals in the dataset?

```{r markerset_dist}
dist_purity <- interIndividualDist(numgen[markers_purity,])

hist(dist_purity, xlab = "Euclidian distance",
     main = "Histogram of inter-individual distances using Purity marker subset")

mean(dist_purity == 0)

dist_simanneal <- interIndividualDist(numgen[markers_simanneal,])

hist(dist_simanneal, xlab = "Euclidian distance",
     main = "Histogram of inter-individual distances using Purity marker subset")

mean(dist_simanneal == 0)
```

If two individuals were selected at random, there is a 0.4% chance that they
would have identical genotypes across all markers using the simulated annealing
algorithm, as opposed to a 0.02% chance using markers from the Purity algorithm.
So, the Purity algorithm is somewhat better for distinguishing individuals.

We can also compare the geometric mean of expected heterozygosity in all of the
populations.

```{r markerset_div}
DivScore(grp_alfreq[markers_purity,])
DivScore(grp_alfreq[markers_simanneal,])
```

The simulated annealing algorithm gives higher expected heterozygosity.

We can also see the geometric mean of Jost's D, a differentiation statistic
among populations.

```{r  markerset_diff}
DiffScore(grp_alfreq[markers_purity,])
DiffScore(grp_alfreq[markers_simanneal,])
```

The simulated annealing algorithm gives higher differentiation among populations.

## Exporting markers

Now we can get the flanking sequences for the markers that we selected.

```{r formatkasp}
markerseq1 <- formatKasp(myvcf, markers_purity, refgenome)
markerseq2 <- formatKasp(myvcf, markers_simanneal, refgenome)

head(markerseq1)
```

These can be exported to a spreadsheet.

```{r export}
write.csv(markerseq1, file = "results/pedigree_verification_markers_Purity.csv",
          row.names = FALSE)
write.csv(markerseq2, file = "results/pedigree_verification_markers_simanneal.csv",
          row.names = FALSE)
```
